cmd/containerd-shim-runhcs-v1/task_hcs.go

```diff
@@ -1008,3 +1026,68 @@ func (ht *hcsTask) requestUpdateContainer(ctx context.Context, resourcePath stri
 	}
 	return ht.c.Modify(ctx, modification)
 }
+
+// publishNotifications monitors the containers notification queue.
+// Currently only OOM events are published.
+//
+// This MUST be called via a goroutine to wait on a background thread.
+func (ht *hcsTask) publishNotifications(ctx context.Context) {
+	var err error
+	ctx, span := trace.StartSpan(ctx, "hcsTask::publishNotifications")
+	defer span.End()
+	defer func() { oc.SetSpanStatus(span, err) }()
+	span.AddAttributes(trace.StringAttribute(logfields.TaskID, ht.id))
+	entity := log.G(ctx).WithField(logfields.TaskID, ht.id)
+
+	nq, err := ht.c.Notifications()
+	if err == notifications.ErrNotificationsNotSupported {
+		entity.Debug("container does not support notifications")
+		return
+	} else if err != nil {
+		entity.WithError(err).Warning("could not access notification queue")
+		return
+	}
+
+	loop := true
+	for loop {
+		select {
+		case <-ctx.Done():
+			loop = false
+		default:
+			v, err := nq.ReadOrWait()
+			if err != nil {
+				if err == queue.ErrQueueClosed {
+					// set err to nil for oc.SetSpanStatus
+					err = nil //nolint:ineffassign
+				} else {
+					// this could be a transient fault, but to prevent infinite
+					// erroneous and polluting the log stream, break here as well
+					entity.WithError(err).Warning("could not read from notification queue")
+				}
+
+				loop = false
+				break
+			}
+
+			ntf, ok := (v).(notifications.Notification)
+			if !ok {
+				err = fmt.Errorf("expected type of 'notifications.Notification{}', received %T", v)
+				entity.WithError(err).Error("improper notification type")
+
+			}
+
+			entity.WithField("event", ntf.String()).Debug("publishing event notifications")
+			if ntf == notifications.Oom {
+				if err := ht.events.publishEvent(
+					ctx,
+					runtime.TaskOOMEventTopic,
+					&eventstypes.TaskOOM{
+						ContainerID: ht.id,
+					}); err != nil {
+					entity.Error("failed to publish TaskOOMEventTopic")
+				}
+			}
```

Comment:
I'd make this a switch statement against the notification type, and the default case is the "how the heck did we get this" scenario. Easier to extend if we add support for other notifs also.

Reply:
Added!